Fontique - Compile error for 32bit

Hey

I'm trying to compile a Slint App for a bit of an exotic architecture (`mipsel-unknown-linux-gnu` - 32 bit LE) with `cross` and I'm getting a compile error for `fontique`. Although Slint depends on `v0.10.0`, this should still apply for `v.0.11.0`. It seems that `self.shared_version` fails to convert and needs specific handling.

From https://docs.rs/atomic-counter/latest/atomic_counter/

> Both implementations are lock-free. Both are a very thin layer over [AtomicUsize](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html) which is more powerful but might be harder to use correctly.

For reference:
https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/std/primitive.usize.html



Let me know if you need more information.

**Build command:**

```sh
cross +nightly build \
    -Z build-std=core,alloc,std,proc_macro,panic_abort \
    --target mipsel-unknown-linux-gnu \
    --release
```

**Error**

```sh
error[E0308]: mismatched types
    --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/fontique-0.10.0/src/collection/mod.rs:268:45
     |
 268 |                 version: AtomicCounter::new(self.shared_version),
     |                          ------------------ ^^^^^^^^^^^^^^^^^^^ expected `u32`, found `u64`
     |                          |
     |                          arguments to this function are incorrect
     |
note: associated function defined here
    --> /rust/lib/rustlib/src/rust/library/core/src/sync/atomic.rs:2623:26
     |
2623 |               pub const fn new(v: $int_type) -> Self {
     |                            ^^^
...
3707 | / atomic_int! {
3708 | |     cfg(target_has_atomic = "32"),
3709 | |     cfg(target_has_atomic_primitive_alignment = "32"),
3710 | |     stable(feature = "integer_atomics_stable", since = "1.34.0"),
...    |
3722 | |     u32 AtomicU32
3723 | | }
     | |_- in this macro invocation
     = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)
help: you can convert a `u64` to a `u32` and panic if the converted value doesn't fit
     |
 268 |                 version: AtomicCounter::new(self.shared_version.try_into().unwrap()),
     |                                                                ++++++++++++++++++++

error[E0308]: mismatched types
   --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/fontique-0.10.0/src/collection/mod.rs:571:39
    |
571 |             if self.shared_version != version {
    |                -------------------    ^^^^^^^ expected `u64`, found `u32`
    |                |
    |                expected because this is `u64`
    |
help: you can convert a `u32` to a `u64`
    |
571 |             if self.shared_version != version.into() {
    |                                              +++++++

error[E0308]: mismatched types
   --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/fontique-0.10.0/src/collection/mod.rs:577:39
    |
577 |                 self.shared_version = version;
    |                 -------------------   ^^^^^^^ expected `u64`, found `u32`
    |                 |
    |                 expected due to the type of this binding
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -119,7 +119,7 @@ jobs:
         uses: dtolnay/rust-toolchain@master
         with:
           toolchain: ${{ env.RUST_STABLE_VER }}
-          targets: x86_64-unknown-none,thumbv8m.main-none-eabihf
+          targets: x86_64-unknown-none,thumbv8m.main-none-eabihf,powerpc-unknown-linux-gnu
           components: clippy
 
       - name: Install fontconfig-dev
@@ -143,6 +143,14 @@ jobs:
 
       - name: cargo clippy (no_std 32-bit)
         run: cargo hack clippy ${{ env.RUST_MIN_VER_PKGS }} --locked --optional-deps --each-feature --ignore-unknown-features --features libm --exclude-features ${{ env.FEATURES_DEPENDING_ON_STD }} --target thumbv8m.main-none-eabihf -- -D warnings
+
+      # We check the std code paths on a 32-bit target without 64-bit atomics
+      # (powerpc-unknown-linux-gnu) to catch regressions like our atomics-based
+      # version counters assuming a 64-bit-wide integer. We omit --optional-deps
+      # and exclude the system feature because the system font backends pull in
+      # C dependencies (e.g. fontconfig) that can't be cross-compiled here.
+      - name: cargo clippy (32-bit with std)
+        run: cargo hack clippy ${{ env.RUST_MIN_VER_PKGS }} --locked --each-feature --ignore-unknown-features --features std --exclude-features default,system,png,vello_hybrid --target powerpc-unknown-linux-gnu -- -D warnings
 
       - name: cargo clippy
         run: cargo hack clippy --workspace --locked --optional-deps --each-feature --ignore-unknown-features --features std -- -D warnings
@@ -275,7 +283,7 @@ jobs:
         uses: dtolnay/rust-toolchain@master
         with:
           toolchain: ${{ env.RUST_MIN_VER }}
-          targets: x86_64-unknown-none,thumbv8m.main-none-eabihf
+          targets: x86_64-unknown-none,thumbv8m.main-none-eabihf,powerpc-unknown-linux-gnu
 
       - name: Install fontconfig-dev
         if: runner.os == 'Linux'
@@ -298,6 +306,14 @@ jobs:
 
       - name: cargo check (no_std 32-bit)
         run: cargo hack check ${{ env.RUST_MIN_VER_PKGS }} --locked --optional-deps --each-feature --ignore-unknown-features --features libm --exclude-features ${{ env.FEATURES_DEPENDING_ON_STD }} --target thumbv8m.main-none-eabihf
+
+      # We check the std code paths on a 32-bit target without 64-bit atomics
+      # (powerpc-unknown-linux-gnu) to catch regressions like our atomics-based
+      # version counters assuming a 64-bit-wide integer. We omit --optional-deps
+      # and exclude the system feature because the system font backends pull in
+      # C dependencies (e.g. fontconfig) that can't be cross-compiled here.
+      - name: cargo check (32-bit with std)
+        run: cargo hack check ${{ env.RUST_MIN_VER_PKGS }} --locked --each-feature --ignore-unknown-features --features std --exclude-features default,system,png,vello_hybrid --target powerpc-unknown-linux-gnu
 
       - name: cargo check
         run: cargo hack check ${{ env.RUST_MIN_VER_PKGS }} --locked --optional-deps --each-feature --ignore-unknown-features --features std

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,12 @@ This release has an [MSRV] of 1.88.
 
 - Breaking change: the `Glyph::style_index` field was removed. Use `Cluster::{style, style_index}` or `GlyphRun::{style, style_index}` instead. (#661 by [@tomcur][])
 
+### Fixed
+
+#### Fontique
+
+- Fix compilation on 32-bit platforms without 64-bit atomics (e.g. `mipsel-unknown-linux-gnu`). (#671 by [@nicoburns][])
+
 ## [0.11.0] - 2026-06-24
 
 This release has an [MSRV] of 1.88.

--- a/fontique/src/collection/mod.rs
+++ b/fontique/src/collection/mod.rs
@@ -21,7 +21,7 @@ use super::{
     generic::GenericFamilyMap,
     source::{SourceId, SourceInfo, SourceKind},
 };
-use crate::AtomicCounter;
+use crate::{AtomicCounter, CounterInt};
 use alloc::{string::String, sync::Arc, vec::Vec};
 use hashbrown::HashMap;
 use read_fonts::types::NameId;
@@ -242,7 +242,7 @@ struct Inner {
     #[allow(unused)]
     shared: Option<Arc<Shared>>,
     #[allow(unused)]
-    shared_version: u64,
+    shared_version: CounterInt,
     fallback_cache: FallbackCache,
 }
 

--- a/fontique/src/lib.rs
+++ b/fontique/src/lib.rs
@@ -75,3 +75,13 @@ pub use source_cache::{SourceCache, SourceCacheOptions};
 use core::sync::atomic::AtomicU32 as AtomicCounter;
 #[cfg(target_has_atomic = "64")]
 use core::sync::atomic::AtomicU64 as AtomicCounter;
+
+/// The integer type underlying [`AtomicCounter`].
+///
+/// This is `u32` on platforms without 64-bit atomics (e.g. many 32-bit
+/// targets) and `u64` elsewhere.
+#[cfg(not(target_has_atomic = "64"))]
+type CounterInt = u32;
+/// The integer type underlying [`AtomicCounter`].
+#[cfg(target_has_atomic = "64")]
+type CounterInt = u64;

